manim/scene/scene.py

```diff
@@ -849,8 +883,36 @@ def get_time_progression(
         if self.renderer.skip_animations and not override_skip_animations:
             times = [run_time]
         else:
-            step = 1 / config["frame_rate"]
-            times = np.arange(0, run_time, step)
+            if speedinfo is not None:
+                nodes = list(speedinfo.keys())
+                speeds = list(speedinfo.values())
+                if nodes[0] != 0:
+                    nodes.insert(0, 0)
+                if nodes[-1] != 1:
+                    nodes.append(1)
+                    speeds.append(speeds[-1])
+                times = np.array([])
+                on_node = 0
+
+                for dur, s in zip(np.diff(nodes), speeds):
+                    rel_speed_change = s / self.speed
+                    adj_func = np.vectorize(
+                        lambda x: (rel_speed_change ** 2 - 1) / 4 * x ** 2 + x,
+                    )
+                    func_unity_at = 2 / (rel_speed_change + 1)
+                    adj_time = run_time * dur
+                    dt = self.speed / config["frame_rate"]
+                    times = np.append(
+                        times,
+                        adj_func(np.arange(0, func_unity_at * adj_time, dt) / adj_time)
+                        * adj_time
+                        + nodes[on_node] * run_time,
+                    )
+                    self.speed = s
+                    on_node += 1
```

Comment:
I get the feeling that it would be cleaner to start with some `previous_node = 0`, then iterating over the dictionary via something like `for node, speed in sorted(speedinfo.items())` and computing the duration as `duration = node - previous_node`, while at the end of the loop setting `previous_node` to `node`.